### PR TITLE
OTLP exporter: Let final retry error include last retryable error message

### PR DESCRIPTION
**Pull request by jmacd**
This improves the error messages printed when an OTel-Go SDK experiences retryable errors that eventually time out. Significantly, this includes text from the final retryable error to help the user understand the why the timeout has been reached.

I [briefly updates the OTel collector example main() to avoid blocking in dial]([be228e5](https://github.com/open-telemetry/opentelemetry-go/pull/3514/commits/be228e547758c64f01b9874030e5b7f0468f0ac4)), to demonstrate the problem. After this change:

```
2022/12/05 11:14:10 context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4318: connect: connection refused"
2022/12/05 11:14:20 context deadline exceeded: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:4318: connect: connection refused"
```

Before this change:

```
2022/12/05 11:14:10 context deadline exceeded
2022/12/05 11:14:20 context deadline exceeded
```

Part of #3513.

### CODE REVIEW

**Review by dashpole — APPROVED** [review]
lgtm, but probably needs a changelog

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,11 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - The `NewMetricExporter` in `go.opentelemetry.io/otel/bridge/opencensus` is deprecated.  Use `NewMetricProducer` instead. (#3541)
 
+### Changed
```
Comment (MrAlias) [inline CHANGELOG.md:32]: This duplicates the `### Changed` section. This needs to be merged there.

File: CHANGELOG.md [hunk 2]
```diff
@@ -29,6 +29,11 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
...
+
+- OTLP exporter for traces and metrics will print the final retryable error message
+  when attempts to retry time out. (#3514)
```
Comment (MrAlias) [inline CHANGELOG.md:35]:
```suggestion
- The OTLP exporters for traces and metrics will print the final retryable error message when attempts to retry time out. (#3514)
```

File: exporters/otlp/internal/retry/retry.go [hunk 3]
```diff
@@ -114,12 +114,12 @@ func (c Config) RequestFunc(evaluate EvaluateFunc) RequestFunc {
...
 				elapsed := b.GetElapsedTime()
 				if b.MaxElapsedTime != 0 && elapsed+throttle > b.MaxElapsedTime {
-					return fmt.Errorf("max retry time would elapse: %w", err)
+					return fmt.Errorf("max retry time would elapse: %w", fnErr)
 				}
 				delay = throttle
 			}
 
-			if err := waitFunc(ctx, delay); err != nil {
+			if err := waitFunc(ctx, fnErr, delay); err != nil {
```
Comment (MrAlias) [inline exporters/otlp/internal/retry/retry.go:122]: Keeping the error formatting in this function alone would ensure that any error returned from `waitFunc` will include the retryable error. That way when another error is returned from `wait` it won't need to duplicate the wrapping.

```suggestion
			if err := waitFunc(ctx, delay); err != nil {
				err = fmt.Errorf("%w: %s", err, fnErr)
```

File: exporters/otlp/internal/retry/retry_test.go [hunk 4]
```diff
@@ -53,13 +52,21 @@ func TestWait(t *testing.T) {
...
 
+	retryable := status.Error(codes.Unavailable, "sending to xyz: connection refused")
 	for _, test := range tests {
-		assert.Equal(t, test.expected, wait(test.ctx, test.delay))
+		err := wait(test.ctx, retryable, test.delay)
+		if test.expectedToBe == nil {
+			assert.NoError(t, err)
+		} else {
+			assert.Error(t, err)
+			assert.True(t, errors.Is(err, test.expectedToBe))
```
Comment (MrAlias) [inline exporters/otlp/internal/retry/retry_test.go:67]: [`assert.ErrorIs`](https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorIs)

File: exporters/otlp/internal/retry/retry_test.go [hunk 5]
```diff
@@ -23,28 +23,27 @@ import (
...
 )
 
 func TestWait(t *testing.T) {
 	tests := []struct {
-		ctx      context.Context
-		delay    time.Duration
-		expected error
+		ctx          context.Context
+		delay        time.Duration
+		expectedToBe error
```
Comment (MrAlias) [inline exporters/otlp/internal/retry/retry_test.go:34]: nit: This rename doesn't seem needed.

**Review by hanyuancheung — APPROVED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -27,11 +27,13 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - `traceIDRatioSampler` (given by `TraceIDRatioBased(float64)`) now uses the rightmost bits for sampling decisions,
   fixing random sampling when using ID generators like `xray.IDGenerator`
   and increasing parity with other language implementations. (#3557)
+- The OTLP exporters for traces and metrics will print the final retryable error message when attempts to retry time out. (#3514)
 
 ### Deprecated
 
 - The `NewMetricExporter` in `go.opentelemetry.io/otel/bridge/opencensus` is deprecated.  Use `NewMetricProducer` instead. (#3541)
 
+
```
Comment (MrAlias) [inline CHANGELOG.md:36]:
```suggestion
```

File: exporters/otlp/internal/retry/retry.go [hunk 7]
```diff
@@ -141,6 +144,7 @@ func wait(ctx context.Context, delay time.Duration) error {
 		select {
 		case <-timer.C:
 		default:
+			// Include the final retryable error as an explanation.
```
Comment (MrAlias) [inline exporters/otlp/internal/retry/retry.go:147]:
```suggestion
```

File: exporters/otlp/internal/retry/retry_test.go [hunk 8]
```diff
@@ -59,7 +56,13 @@ func TestWait(t *testing.T) {
 	}
 
 	for _, test := range tests {
-		assert.Equal(t, test.expected, wait(test.ctx, test.delay))
+		err := wait(test.ctx, test.delay)
+		if test.expected == nil {
+			assert.NoError(t, err)
+		} else {
+			assert.Error(t, err)
```
Comment (MrAlias) [inline exporters/otlp/internal/retry/retry_test.go:63]:
```suggestion
```

**Review by MrAlias — APPROVED** [review]

File: exporters/otlp/internal/retry/retry_test.go (hunk 5 above)
Comment (jmacd) [reply exporters/otlp/internal/retry/retry_test.go]: Yes, I was thinking that the condition had changed from expected equality to expected "errors.Is" (meaning "to be").

File: exporters/otlp/internal/retry/retry_test.go (hunk 4 above)
Comment (jmacd) [reply exporters/otlp/internal/retry/retry_test.go]: Neat trick!

File: CHANGELOG.md (hunk 1 above)
Comment (jmacd) [reply CHANGELOG.md]: I had been confused by that. It turns out it's the automatic merge that was creating the duplicate section, so I couldn't reproduce it locally.

**Comment by jmacd** [comment]
Thanks for the help @MadVikingGod.

**Pull request merged**